src/exchange/exchange.py

```diff
@@ -72,12 +75,36 @@ def generate(self) -> Message:
         """Generate the next message."""
         self.moderator.rewrite(self)
 
-        message, usage = self.provider.complete(
-            self.model,
-            self.system,
-            messages=self.messages,
-            tools=self.tools,
-        )
+        num_times_attempted = 0
+        while num_times_attempted < 3:
```

Comment:
current retry in the provider is 5. does this mean we will retry 15 times with worst scenarios?

Reply:
Yea, so here we are retrying after modifying the messages array. We try popping off as little context as possible, and can try doing that up to 3 times. I am not hard-set on the `3` times, but figured it would at least land us in a spot for subsequent discussion with the team.